### PR TITLE
Introduce RecordContext and Base components for Edit, Create and Show

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * Use the useRecordContext() hook to read the context. That's what the Edit components do in react-admn.
+ *
+ * @example
+ *
+ * import { useEditController, EditContext } from 'ra-core';
+ *
+ * const Edit = props => {
+ *     const controllerProps = useEditController(props);
+ *     return (
+ *         <EditContext.Provider value={controllerProps}>
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:23]: should be RecordContext

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import { createContext, useContext } from 'react';
+import pick from 'lodash/pick';
+import { Record } from '../types';
+
+export interface RecordContextValue {
+    record?: Partial<Record>;
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:6]: why partial? the id is required

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ * Must be used within a <EditContextProvider> (e.g. as a descendent of <Edit>
+ * or <EditBase>) or within a <ShowContextProvider> (e.g. as a descendent of <Show>
+ * or <ShowBase>)
+ *
+ * @returns {RecordContextValue} The record context
+ */
+export const useRecordContext = () => {
+    const context = useContext<RecordContextValue>(RecordContext);
+
+    if (!context) {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:57]: This will never happen. If you call `useContext` outside of the Context.Provider, you'll get the default context (in your case, the context line 29).

```jsx
import React,  { createContext, useContext } from "react";

const FooContext = createContext({ hello: 'world'})

export default function App() {
  const context = useContext(FooContext);
  return (
      <h1>Hello {context.hello}</h1>
  );
}

// always render "Hello world"
```

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * };
+ */
+export const RecordContext = createContext<RecordContextValue>({
+    loaded: null,
+    loading: null,
+    record: null,
+});
+
+RecordContext.displayName = 'RecordContext';
+
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:36]: you should also expose a RecordContextProvider, to let us change the internal storage in the future without breaking compatibility

File: packages/ra-core/src/controller/SaveContext.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * const {
+ *     save,
+ *     saving
+ * } = useSaveContext();
+ */
+export const useSaveContext = () => {
+    const context = useContext(SaveContext);
+
+    if (!context) {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/SaveContext.tsx:46]: same, this will never happen

File: packages/ra-core/src/controller/details/CreateContext.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+CreateContext.displayName = 'CreateContext';
+
+export const useCreateContext = <
+    RecordType extends Record = Record
+>(): CreateControllerProps<RecordType> => {
+    const context = useContext(CreateContext);
+
+    // @ts-ignore
+    return context;
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/CreateContext.tsx:48]: why is this necessary? Please expand the comment

File: packages/ra-core/src/controller/details/CreateContextProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ */
+export const CreateContextProvider = ({
+    children,
+    value,
+}: {
+    children: ReactElement;
+    value: CreateControllerProps;
+}) => (
+    <CreateContext.Provider value={value}>
+        <SideEffectContext.Provider value={usePickSideEffectContext(value)}>
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/CreateContextProvider.tsx:40]: Shouldn't we merge the SideEffectContext and the SaveContext?

File: packages/ra-core/src/controller/details/EditContext.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * const Edit = props => {
+ *     const controllerProps = useEditController(props);
+ *     return (
+ *         <EditContext.Provider value={controllerProps}>
+ *             ...
+ *         </EditContext.Provider>
+ *     );
+ * };
+ */
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/EditContext.tsx:22]: you should probably document the preference for EditContextProvider

File: packages/ra-core/src/controller/details/EditContextProvider.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
+import * as React from 'react';
+import { ReactElement } from 'react';
+import { RecordContext, usePickRecordContext } from '../RecordContext';
+import { EditContext } from './EditContext';
+import { EditControllerProps } from './useEditController';
+import { SideEffectContext, usePickSideEffectContext } from '../saveModifiers';
+import { SaveContext, usePickSaveContext } from '../SaveContext';
+
+/**
+ * Create a Edit Context.
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/EditContextProvider.tsx:10]:
```suggestion
 * Create an Edit Context.
```

File: packages/ra-core/src/controller/details/ShowContext.tsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+ * import { useShowController, ShowContext } from 'ra-core';
+ *
+ * const Show = props => {
+ *     const controllerProps = useShowController(props);
+ *     return (
+ *         <ShowContext.Provider value={controllerProps}>
+ *             ...
+ *         </ShowContext.Provider>
+ *     );
+ * };
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/ShowContext.tsx:21]: same, we should document the preference for ShowContextProvider

File: packages/ra-core/src/controller/details/ShowContextProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import * as React from 'react';
+import { ReactElement } from 'react';
+import { RecordContext, usePickRecordContext } from '../RecordContext';
+import { ShowContext } from './ShowContext';
+import { ShowControllerProps } from './useShowController';
+
+/**
+ * Show a Show Context.
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/ShowContextProvider.tsx:8]:
```suggestion
 * Create a Show Context.
```

File: packages/ra-core/src/controller/details/ShowContextProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+ */
+export const ShowContextProvider = ({
+    children,
+    value,
+}: {
+    children: ReactElement;
+    value: ShowControllerProps;
+}) => (
+    <ShowContext.Provider value={value}>
+        <RecordContext.Provider value={usePickRecordContext(value)}>
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/ShowContextProvider.tsx:36]: RecordContextProvider?

File: packages/ra-core/src/controller/details/index.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+export * from './EditBase';
+export * from './EditContext';
+export * from './EditContextProvider';
+export * from './EditController';
+export * from './ShowBase';
+export * from './ShowContext';
+export * from './ShowContextProvider';
+export * from './ShowController';
+
+// We don't want to export CreateProps, EditProps and ShowProps as they should
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/index.ts:21]: I think they're exported anyway because of the export * above

File: packages/ra-ui-materialui/src/detail/Create.tsx [hunk 14]
```diff
@@ -1,18 +1,13 @@
...
+    CreateContextProvider,
     useCheckMinimumRequiredProps,
     useCreateController,
-    SideEffectContext,
-    CreateControllerProps,
 } from 'ra-core';
 
-import TitleForRecord from '../layout/TitleForRecord';
 import { CreateProps } from '../types';
+import { CreateView } from '.';
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/Create.tsx:10]: I don't like these, they can create circular preferences. Prefer importing from the file is it's in the same directory.

File: packages/ra-ui-materialui/src/detail/CreateActions.js [hunk 15]
```diff
@@ -37,11 +38,14 @@ const sanitizeRestProps = ({
  *         </Create>
  *     );
  */
-const CreateActions = ({ basePath, className, hasList, ...rest }) => (
-    <TopToolbar className={className} {...sanitizeRestProps(rest)}>
-        {hasList && <ListButton basePath={basePath} />}
-    </TopToolbar>
-);
+const CreateActions = ({ className, ...rest }) => {
+    const { basePath, hasList } = useCreateContext();
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/CreateActions.js:42]: you should pass the component props, to provide a fallback in case the context isn't defined (for BC)

File: packages/ra-core/src/controller/details/CreateContext.tsx [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    saving: null,
+    successMessage: null,
+    version: null,
+});
+
+CreateContext.displayName = 'CreateContext';
+
+export const useCreateContext = <
+    RecordType extends Record = Record
+>(): CreateControllerProps<RecordType> => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/CreateContext.tsx:44]: See what I did in useListContext: if the context is empty, we fallback to props. This allows for a smooth transition.

```
const useListContext = <RecordType extends Record = Record>(
    props?: any
): ListControllerProps<RecordType> => {
    const context = useContext(ListContext);
    if (!context.resource) {
        /**
         * The element isn't inside a <ListContext.Provider>
         *
         * This may only happen when using Datagrid / SimpleList / SingleFieldList components
         * outside of a List / ReferenceManyField / ReferenceArrayField -
         * which isn't documented but tolerated.
         * To avoid breakage in that case, fallback to props
         *
         * @deprecated - to be removed in 4.0
         */
        if (process.env.NODE_ENV !== 'production') {
            console.log(
                "List components must be used inside a <ListContext.Provider>. Relying on props rather than context to get List data and callbacks is deprecated and won't be supported in the next major version of react-admin."
            );
        }
        return props;
    }
    // @ts-ignore
    return context;
};
```

File: packages/ra-core/src/controller/details/CreateContext.tsx [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    resource: null,
+    save: null,
+    saving: null,
+    successMessage: null,
+    version: null,
+});
+
+CreateContext.displayName = 'CreateContext';
+
+export const useCreateContext = <
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/CreateContext.tsx:42]: I think this one deserves its own file, to be easier to find in IDEs

File: packages/ra-core/src/controller/details/EditContext.tsx [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    saving: null,
+    successMessage: null,
+    version: null,
+});
+
+EditContext.displayName = 'EditContext';
+
+export const useEditContext = <
+    RecordType extends Record = Record
+>(): EditControllerProps<RecordType> => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/EditContext.tsx:44]: Same remark as for the useEditContext: fallback to props for BC, extract to another file

File: packages/ra-core/src/controller/details/ShowContext.tsx [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    defaultTitle: null,
+    loaded: null,
+    loading: null,
+    resource: null,
+    version: null,
+});
+
+ShowContext.displayName = 'ShowContext';
+
+export const useShowContext = <
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/ShowContext.tsx:35]: Same remark as for the useEditContext: fallback to props for BC, extract to another file

File: packages/ra-ui-materialui/src/detail/CreateView.tsx [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+        resource,
+        save,
+        setOnSuccess,
+        setOnFailure,
+        setTransform,
+        saving,
+        title,
+        version,
+        ...rest
+    } = props;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/CreateView.tsx:34]: CreateView should use the CreateContext to get the controller props. See how I did in ListView

File: packages/ra-ui-materialui/src/detail/EditActions.tsx [hunk 21]
```diff
@@ -30,18 +30,15 @@ import TopToolbar from '../layout/TopToolbar';
...
-    hasShow,
-    hasList,
-    ...rest
-}: EditActionsProps) => (
-    <TopToolbar className={className} {...rest}>
-        {hasShow && <ShowButton basePath={basePath} record={data} />}
-    </TopToolbar>
-);
+const EditActions = ({ className, ...rest }: EditActionsProps) => {
+    const { basePath, hasShow, record: data } = useEditContext();
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/EditActions.tsx:34]: pass the props to useEditContext to allow a smooth transition

File: packages/ra-ui-materialui/src/detail/EditView.tsx [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+        save,
+        setOnSuccess,
+        setOnFailure,
+        setTransform,
+        saving,
+        title,
+        undoable,
+        version,
+        ...rest
+    } = props;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/EditView.tsx:42]: should use useEditContext to get controller props

File: packages/ra-ui-materialui/src/detail/ShowActions.tsx [hunk 23]
```diff
@@ -40,17 +40,14 @@ const sanitizeRestProps = ({
...
-    data,
-    hasEdit,
-    ...rest
-}: ShowActionsProps) => (
-    <TopToolbar className={className} {...sanitizeRestProps(rest)}>
-        {hasEdit && <EditButton basePath={basePath} record={data} />}
-    </TopToolbar>
-);
+const ShowActions = ({ className, ...rest }: ShowActionsProps) => {
+    const { basePath, hasEdit, record: data } = useShowContext();
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/ShowActions.tsx:44]: pass props to the useShowContext call

File: packages/ra-ui-materialui/src/detail/ShowView.tsx [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+        component: Content,
+        defaultTitle,
+        hasEdit,
+        hasList,
+        record,
+        resource,
+        title,
+        version,
+        ...rest
+    } = props;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/ShowView.tsx:34]: grab the show controller props from useShowContext

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Awesome! This will ease Edit, Create and Show views customization big time.

File: packages/ra-core/src/controller/RecordContext.tsx (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/controller/RecordContext.tsx:6]: Not when passing a record for creation

File: packages/ra-core/src/controller/RecordContext.tsx (hunk 4 above)
Comment (djhi) [reply packages/ra-core/src/controller/RecordContext.tsx:36]: Did it locally already

File: packages/ra-core/src/controller/details/CreateContextProvider.tsx (hunk 7 above)
Comment (djhi) [reply packages/ra-core/src/controller/details/CreateContextProvider.tsx:40]: ongoing

File: packages/ra-core/src/controller/details/index.ts (hunk 13 above)
Comment (djhi) [reply packages/ra-core/src/controller/details/index.ts:21]: They're not. Otherwise TS would have complained

Comment (fzaninotto) [reply packages/ra-core/src/controller/details/index.ts:21]: hmm I wouldn't trust TypeScript on this one

Comment (djhi) [reply packages/ra-core/src/controller/details/index.ts:21]: And no, because they are declared in the files for which we don't have an `export *`

File: packages/ra-ui-materialui/src/button/SaveButton.tsx [hunk 25]
```diff
@@ -106,6 +104,7 @@ const SaveButton: FC<SaveButtonProps> = props => {
             setOnSave();
         }
         if (onSuccess) {
+            console.log('onSuccess');
```
Comment (Luwangel) [inline packages/ra-ui-materialui/src/button/SaveButton.tsx:107]:
```suggestion
```

**Review by Luwangel — APPROVED** [review]
So many changes. Well done!

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+ * Context to store the result of the useRecord() hook.
+ *
+ * Use the useRecordContext() hook to read the context. That's what the Edit and Show components do in react-admn.
+ *
+ * @example
+ *
+ * import { useEditController, EditContext } from 'ra-core';
+ *
+ * const Edit = props => {
+ *     const controllerProps = useEditController(props);
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:16]: bad code example. Shouldn't it be:

```
const { record }= useEditController(props);
return (
     <RecordContextProvider value={record}>
        ...
    </RecordContextProvider>
);

??

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+ *     return (
+ *         <RecordContextProvider value={controllerProps}>
+ *             ...
+ *         </RecordContextProvider>
+ *     );
+ * };
+ */
+export const RecordContext = createContext<Record | Omit<Record, 'id'>>(
+    undefined
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:25]: no need for undefined, just pass no value

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+export const usePickRecordContext = <
+    RecordType extends Record | Omit<Record, 'id'> = Record
+>(
+    context: RecordType
+) => {
+    return pick(context, ['record']);
+};
+
+/**
+ * Hook to read the record from a context which provide one, such as the EditContext or ShowContext.
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:45]:
```suggestion
 * Hook to read the record from a RecordContext.
```

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+
+export const RecordContextProvider = ({
+    children,
+    value,
+}: RecordContextOptions) => (
+    <RecordContext.Provider value={value}>{children}</RecordContext.Provider>
+);
+RecordContext.displayName = 'RecordContext';
+
+export const usePickRecordContext = <
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:36]: I'd call it `pickRecord` because:

- it's not a hook
- it can pick record from anything, not just a context

Same for other pickers by the way

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+    RecordType extends Record | Omit<Record, 'id'> = Record
+>(
+    context: RecordType
+) => {
+    return pick(context, ['record']);
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:41]: I'm thinking: every time something changes in a context (e.g. the selected ids in the list), the pick command will return a new instance of the record, re-rendering everything. 

I think it's not a premature optimization to memoize the picking.

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * Must be used within a <RecordContext> such as provided by the <EditContextProvider>
+ * (e.g. as a descendent of <Edit> or <EditBase>) or within a <ShowContextProvider>
+ * (e.g. as a descendent of <Show> or <ShowBase>)
+ *
+ * @returns {Record} The record context
+ */
+export const useRecordContext = <
+    RecordType extends Record | Omit<Record, 'id'> = Record
+>() => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:55]: you must expect a record to be passed here to allow backwards compatibility

File: packages/ra-core/src/controller/details/SaveContext.tsx [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+ * Used in forms.
+ *
+ * @example
+ *
+ * const {
+ *     save,
+ *     saving
+ * } = useSaveContext();
+ */
+export const useSaveContext = () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/SaveContext.tsx:48]: must accept props as params for BC

File: packages/ra-core/src/controller/details/SaveContext.tsx [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+    const context = useContext(SaveContext);
+
+    if (!context) {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/details/SaveContext.tsx:51]: this will never happen: the default context is an object

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Almost there!

File: packages/ra-core/src/controller/RecordContext.tsx (hunk 27 above)
Comment (djhi) [reply packages/ra-core/src/controller/RecordContext.tsx:25]: TS disagree

File: packages/ra-core/src/controller/RecordContext.tsx (hunk 29 above)
Comment (djhi) [reply packages/ra-core/src/controller/RecordContext.tsx:36]: Should I do the same for all pickers which were already there for the ListContextProvider ?

Comment (fzaninotto) [reply packages/ra-core/src/controller/RecordContext.tsx:36]: yep

Comment (djhi) [reply packages/ra-core/src/controller/RecordContext.tsx:36]: Not possible as they use hooks

File: examples/simple/src/comments/CommentEdit.js [hunk 34]
```diff
@@ -43,71 +44,76 @@ const inputText = record => `${record.title} - ${record.id}`;
...
-                        version={version}
-                    >
-                        <TextInput disabled source="id" fullWidth />
-                        <ReferenceInput
-                            source="post_id"
-                            reference="posts"
-                            perPage={15}
-                            sort={{ field: 'title', order: 'ASC' }}
-                            fullWidth
+                        data={record}
```
Comment (fzaninotto) [inline examples/simple/src/comments/CommentEdit.js:65]: shouldn't be required, since it's in the Edit context

File: examples/simple/src/comments/CommentEdit.js [hunk 35]
```diff
@@ -43,71 +44,76 @@ const inputText = record => `${record.title} - ${record.id}`;
...
+                        hasList
+                    />
+                </div>
+                <Card className={classes.card}>
+                    {record && (
+                        <SimpleForm
+                            basePath={basePath}
+                            redirect={redirect}
+                            resource={resource}
+                            record={record}
```
Comment (fzaninotto) [inline examples/simple/src/comments/CommentEdit.js:76]: same

File: packages/ra-core/src/controller/RecordContext.tsx [hunk 36]
```diff
@@ -0,0 +1,78 @@
+import * as React from 'react';
+import { createContext, ReactNode, useContext, useMemo } from 'react';
+import pick from 'lodash/pick';
+import { Record } from '../types';
+
+/**
+ * Context to store the result of the current record.
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/RecordContext.tsx:7]:
```suggestion
 * Context to store the current record.
```

**Review by fzaninotto — CHANGES_REQUESTED** [review]

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
I'm so happy to have this merged!
